feat: add AI review infrastructure (Phase 0) (#2)
tonyandrewmeyer/charmhub-listing-review: AI_ROADMAP.md

- Add `github-copilot-sdk` as an optional dependency group in `pyproject.toml`:
  `[dependency-groups] ai = ["github-copilot-sdk"]`
- Add new CLI entry points as needed in later phases.

---

## Phase 1: Smart Failure Explanations

**Goal:** When checks fail, generate contextual, actionable fix instructions
instead of generic messages.

**Scope:** Simplest AI integration — validates the full SDK pipeline
end-to-end.

**Implementation:**

- Add `async explain_failures(results: list[CheckResult]) -> list[CheckResult]`
  to `ai_client.py`.
- For each failed result, prompt the LLM with the check name, description, and
  context data.
- Returns 2-3 sentence actionable fix instructions.
- Populate `CheckResult.ai_explanation` with the response.

**Integration:**

- `self_review.py` — display explanations as indented text below each failed
  check.
- `update_issue.py` — append AI explanations as sub-bullets under failed items
  in the issue comment.

**Example output:**

```
 X The charm provides a license statement.
   AI: The LICENSE file was not recognized as a standard open-source license.
   Ensure you are using an unmodified Apache 2.0, GPL 2/3, LGPL 3, or MPL 2.0

>    license text.

Author: ```suggestion
   license text, or open a feature request to recognise other licenses.
```

```

---

## Phase 2: AI-Powered Review Summary

**Goal:** After all checks run, provide a human-readable summary with

> prioritized action items.

Author: ```suggestion
prioritised action items.
```

**Implementation:**

- Add `async generate_summary(charm_name, results, charmcraft_data) -> str`
  to `ai_client.py`.

> - System prompt instructs the LLM to write 3-5 bullet points summarizing

Author: ```suggestion
- System prompt instructs the LLM to write 3-5 bullet points summarising
```

>   readiness for public listing, prioritized by impact.

Author: ```suggestion
  readiness for public listing, prioritised by impact.
```

**Implementation:**

- Add `doc_quality_context(repo_dir) -> dict` to `evaluate.py` — collects
  README.md content (truncated to ~4000 chars), docs/ directory listing,
  documentation URL from charmcraft.yaml.
- Add `async assess_documentation(doc_context) -> str` to `ai_client.py`.
- LLM evaluates: clarity, completeness (installation, configuration, usage,
  troubleshooting), presence of code examples, proper formatting.
- Returns a brief assessment (pass/needs-work/fail) with 2-3 specific
  suggestions.
- The LLM session includes the `read_file` tool so it can request additional
  doc files if needed.

### Phase 3b: Metadata Quality

**Goal:** Evaluate whether `summary`, `description`, `title`, and other text
fields in charmcraft.yaml are well-written and descriptive.

**Implementation:**

- Add `async assess_metadata(charmcraft_data) -> str` to `ai_client.py`.
- LLM evaluates text fields for clarity, informativeness, and best practices.
- Provides specific rewrite suggestions where needed.

**Example output:**

```
AI Metadata Assessment
----------------------
- Summary: Good. Clear and concise.
- Description: Needs work. Too generic — consider mentioning specific

>   integrations (e.g., "integrates with PostgreSQL and S3").

Author: ```suggestion
  integrations (for example, "integrates with PostgreSQL and S3").
```

- Title: "My Charm" should be more descriptive.
```

**Integration (both 3a and 3b):**

- New sections in `self_review.py` output after the main checklist.
- New sections in `update_issue.py` issue comments.

---

## Phase 4: Code Quality Analysis

> **Goal:** Analyze the charm's Python code for antipatterns, missing error

Author: ```suggestion
**Goal:** Analyse the charm's Python code for antipatterns, missing error
```

handling, and Juju/Ops framework misuse.

**Implementation:**

- New module: `ai_code_review.py`
  - `collect_charm_code(repo_dir) -> dict` — finds and reads the main charm
    class, files under `src/`, charm libraries under `lib/`. Truncates each
    file to ~3000 chars. Returns `{filepath: content}`.

>   - `async analyze_code(code_context) -> str` — sends code to Copilot with a

Author: ```suggestion
  - `async analyse_code(code_context) -> str` — sends code to Copilot with a
```

>     specialized system prompt covering:

Author: ```suggestion
    specialised system prompt covering:
```

    1. Common antipatterns (blocking in event handlers, hardcoded values)
    2. Missing error handling (uncaught exceptions in relation handlers)
    3. Ops framework misuse (incorrect status setting, missing `defer()`)
    4. Security concerns (secrets in plain text, unsafe subprocess calls)
  - Returns top 5 findings with severity (info/warning/error) and fix
    suggestions.
  - LLM session includes `read_file` and `list_files` tools for repo
    exploration.

**Integration:**

- `self_review.py` — opt-in via `--code-review` flag (off by default, slower).
  Display results in a "Code Quality" section.
- `update_issue.py` — collapsible `<details>` section in issue comments.

---

## Phase 5: Interactive Self-Review Assistant

**Goal:** Turn `self-review` into a conversational tool where authors can ask
follow-up questions about failures and get guidance.

**Depends on:** All prior phases (uses them as callable tools).

**Implementation:**

- New module: `interactive.py`
  - Async REPL loop maintaining a Copilot session.

>   - Session initialized with: charm name, repository URL, all check results,

Author: ```suggestion
  - Session initialised with: charm name, repository URL, all check results,
```

    and any AI assessments already generated.
  - Custom tools registered via `@define_tool`:
    - `run_check(check_name)` — re-runs a specific check.
    - `read_charm_file(path)` — reads a file from the cloned repo.
    - `explain_check(check_name)` — detailed info about what a check does.
    - `suggest_fix(check_name)` — generates fix suggestion (reuses Phase 1).
    - `assess_docs()` — runs doc quality assessment (reuses Phase 3a).
    - `assess_metadata()` — runs metadata assessment (reuses Phase 3b).
    - `review_code(file_path)` — reviews a specific file (reuses Phase 4).

**Integration:**

- New entry point in `pyproject.toml`:
  `interactive-review = "charmhub_listing_review.interactive:main"`
- `self_review.py` — `--interactive` flag launches the interactive mode after
  showing initial results.

**Example interaction:**

```
$ self-review --charm-name foo --repository https://github.com/org/foo-operator --interactive

[... normal self-review output ...]

Interactive Review Assistant (type 'quit' to exit)

> Why did the license check fail?

> The license file was found but its content hash didn't match any recognized

Author: ```suggestion
The license file was found but its content hash didn't match any recognised
```

license. Your LICENSE file may have been modified from the standard text.

> Can you check if my README has usage examples?
[reads README.md] Your README includes an installation section but lacks usage
examples. Consider adding a "Usage" section showing how to deploy and configure
the charm with `juju deploy`.

> quit
```

> ---

Author: Let's add a phase 6 where we investigate whether it is feasible to use one of the Canonicmal inference snaps for any/all of this functionality. If so then make that an option, either copilot or the snap. We could need to make a wrapper for the snap access (the OpenAI API I think) to slot it in where we have copilot if the user asked for that.
